main.go

```diff
@@ -0,0 +1,124 @@
+// Copyright (c) 2019-2020, Cloudflare, Inc. and Apple, Inc. All rights reserved.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package main
+
+import (
+	"crypto/rand"
+	"fmt"
+	"net/http"
+	"os"
+	"time"
+
+	"github.com/cisco/go-hpke"
+	"github.com/cloudflare/odoh-go"
+	"github.com/jessevdk/go-flags"
+	log "github.com/sirupsen/logrus"
+)
+
+const (
+	// HPKE constants
+	kemID  = hpke.DHKEM_X25519
+	kdfID  = hpke.KDF_HKDF_SHA256
+	aeadID = hpke.AEAD_AESGCM128
+
+	// Keying material (seed) should have as many bits of entropy as the bit length of the x25519 secret key
+	defaultSeedLength = 32
+)
+
+// Set by build process
+var version = "dev"
+
+// CLI flags
+var opts struct {
+	ListenAddr string `short:"l" long:"listen" description:"Address to listen on" default:"localhost:8080"`
+	Resolver   string `short:"r" long:"resolver" description:"Target DNS resolver" default:"127.0.0.1:53"`
+	DisableTls bool   `short:"t" long:"no-tls" description:"Disable TLS"`
+	Cert       string `short:"c" long:"cert" description:"TLS certificate file"`
+	Key        string `short:"k" long:"key" description:"TLS key file"`
+	Verbose    bool   `short:"v" long:"verbose" description:"Enable verbose logging"`
+}
+
+func main() {
+	// Parse cli flags
+	_, err := flags.ParseArgs(&opts, os.Args)
+	if err != nil {
+		os.Exit(1)
+	}
+
+	// Enable debug logging in development releases
+	if //noinspection GoBoolExpressions
+	version == "devel" || opts.Verbose {
+		log.SetLevel(log.DebugLevel)
+		log.Debugln("Verbose logging enabled")
+	}
+
+	// Validate TLS cert/key
+	if !opts.DisableTls && (opts.Cert == "" || opts.Key == "") {
+		log.Fatal("--cert and --key must be set when TLS is enabled")
+	}
+
+	// Random seed for HPKE keypair
+	seed := make([]byte, defaultSeedLength)
+	_, err = rand.Read(seed)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	keyPair, err := odoh.CreateKeyPairFromSeed(hpke.DHKEM_X25519, hpke.KDF_HKDF_SHA256, hpke.AEAD_AESGCM128, seed)
```

Comment:
Is it okay to have the keypair only rotate on process restart? is it worth having the seed periodically roll over within the lifetime of the process?

Reply:
I'll have to look deeper into the IETF draft to see if there's an explicit policy about this. From an initial look I'm not seeing any reference to automated key rollover, so if we were to add this what interval would you recommend? Should it be a configurable parameter?

Also, one thing that we may add later is configuration for long lived TLS sessions that are set up when the process starts. We'd need to keep this in mind during key rollover.

Reply:
i imagine this would be needed for resuming sessions more efficiently, or something of that nature? having it rotate once a day seems like a reasonable default there.
Based on how it's passed in / used, I would expect that long-lived connections would continue to use whatever key existed when they were initiated, right?

Reply:
I'll have to test to make sure, but it seems logical that we would keep any long lived sessions with their original key unless it's required by the protocol to keep it current.